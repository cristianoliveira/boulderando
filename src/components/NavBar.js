import * as React from 'react'
import { useRouter } from 'next/router'

import AppBar from '@mui/material/AppBar'
import Box from '@mui/material/Box'
import Toolbar from '@mui/material/Toolbar'
import Typography from '@mui/material/Typography'
import Button from '@mui/material/Button'
import IconButton from '@mui/material/IconButton'
import MenuIcon from '@mui/icons-material/Menu'
import Menu from '@mui/material/Menu'
import MenuItem from '@mui/material/MenuItem'
import AccountCircle from '@mui/icons-material/AccountCircle'

import { UserContext } from '../context/User'

const PAGE_TITLES = {
  '/': 'Book Session',
  '/sessions': 'Book Session',
  '/user/new': 'New user',
  '/user/edit': 'Editing user',
}

export const NAVBAR_CONTAINER = 'navbar_container'
export const NAVBAR_USER_MENU_BUTTON = 'navbar_user_menu_button'
export const NAVBAR_USER_MENU_CONTAINER = 'navbar_user_menu_button--container'
export const NAVBAR_USER_MENU_ITEM_DELETE = 'navbar_user_menu_button--delete'
export const NAVBAR_USER_MENU_ITEM_EDIT = 'navbar_user_menu_button--edit'

export default function NavBar() {
  const [auth, setAuth] = React.useState(true)
  const [anchorEl, setAnchorEl] = React.useState(null)
  const userContext = React.useContext(UserContext)
  const { user, deleteUser, editUser } = userContext
  const router = useRouter()

  const handleChange = (event) => {
    setAuth(event.target.checked)
  }

  const handleMenu = (event) => {
    setAnchorEl(event.currentTarget)
  }

  const handleClose = () => {
    setAnchorEl(null)
  }

  return (
    <Box sx={{ flexGrow: 1 }}>
      <AppBar position="static" data-testid={NAVBAR_CONTAINER}>
        <Toolbar>
          <Typography variant="h6" component="div" sx={{ flexGrow: 1 }}>
            {PAGE_TITLES[router.asPath]}
          </Typography>
          {user && (
            <div>
              <IconButton
                data-testid={NAVBAR_USER_MENU_BUTTON}
                size="large"
                aria-label="account of current user"
                aria-controls="menu-appbar"
                aria-haspopup="true"
                onClick={handleMenu}
                color="inherit"
              >
                <Box sx={{ m: 1 }}>
                  <AccountCircle />
                </Box>
                <Typography variant="h6" component="small" sx={{ flexGrow: 1 }}>
                  {user.name}
                </Typography>
              </IconButton>
              <Menu
                data-tesid={NAVBAR_USER_MENU_CONTAINER}
                id="menu-appbar"
                anchorEl={anchorEl}
                anchorOrigin={{
                  vertical: 'top',
                  horizontal: 'right',
                }}
                transformOrigin={{
                  vertical: 'top',
                  horizontal: 'right',
                }}
                open={Boolean(anchorEl)}
                onClose={handleClose}
                keepMounted
                autoFocus
              >
                <MenuItem
                  data-testid={NAVBAR_USER_MENU_ITEM_DELETE}
                  onClick={deleteUser}
                >
                  Delete
                </MenuItem>
                <MenuItem
                  data-testid={NAVBAR_USER_MENU_ITEM_EDIT}
                  onClick={editUser}
                >
                  Edit
                </MenuItem>
              </Menu>
            </div>
          )}
        </Toolbar>
      </AppBar>
    </Box>
  )
}
