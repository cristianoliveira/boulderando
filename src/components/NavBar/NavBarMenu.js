import * as React from 'react'
import IconButton from '@mui/material/IconButton'
import Menu from '@mui/material/Menu'
import MenuItem from '@mui/material/MenuItem'
import ListItemText from '@mui/material/ListItemText'
import ListItemIcon from '@mui/material/ListItemIcon'

const ITEM_HEIGHT = 48

export default function LongMenu({ MenuIcon, options, onClick, ...props }) {
  const [selected, setSelected] = React.useState("")
  const [anchorEl, setAnchorEl] = React.useState(null)
  const open = Boolean(anchorEl)
  const handleClick = (event) => {
    setAnchorEl(event.currentTarget)
  }
  const handleClose = () => {
    setAnchorEl(null)
  }

  return (
    <>
      <IconButton
        aria-label="more"
        id="long-button"
        aria-controls={open ? 'long-menu' : undefined}
        aria-expanded={open ? 'true' : undefined}
        aria-haspopup="true"
        onClick={handleClick}
        color="inherit"
        data-testid={props['data-testid']}
      >
        <MenuIcon />
      </IconButton>
      <Menu
        id="long-menu"
        MenuListProps={{
          'aria-labelledby': 'long-button',
        }}
        anchorEl={anchorEl}
        open={open}
        onClose={handleClose}
        PaperProps={{
          style: {
            maxHeight: ITEM_HEIGHT * 4.5,
            width: '20ch',
          },
        }}
      >
        {options.map(({ id, Icon, label, ...oProps }) => (
          <MenuItem
            key={id}
            selected={id === selected}
            onClick={() => {
              setSelected(id);
              onClick({ id, label });
              handleClose()
            }}
            {...oProps}
          >
          {Icon && (
            <ListItemIcon>
              <Icon fontSize="small" />
            </ListItemIcon>
          )}
            <ListItemText>{label}</ListItemText>
          </MenuItem>
        ))}
      </Menu>
    </>
  )
}
