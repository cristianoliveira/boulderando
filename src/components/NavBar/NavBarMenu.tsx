import { useState } from 'react'
import { OverridableComponent } from '@mui/material/OverridableComponent'
import SvgIcon from '@mui/material/SvgIcon'
import IconButton from '@mui/material/IconButton'
import Menu from '@mui/material/Menu'
import MenuItem from '@mui/material/MenuItem'
import ListItemText from '@mui/material/ListItemText'
import ListItemIcon from '@mui/material/ListItemIcon'

const ITEM_HEIGHT = 54

type MenuOptionId = string

type MenuOption = {
  id: MenuOptionId
  label: string
  'data-testid': string
  Icon?: OverridableComponent<any>
  Component?: OverridableComponent<any>
} & Partial<Disableable>

type LongMenuProps = {
  MenuIcon: typeof SvgIcon
  options: MenuOption[]
  'data-testid': string
} & Clickable<Partial<MenuOption>>

export default function LongMenu({
  MenuIcon,
  options,
  onClick,
  'data-testid': dataTestId,
}: LongMenuProps) {
  const [selected, setSelected] = useState<MenuOptionId>('')
  const [anchorEl, setAnchorEl] = useState<HTMLElement | null>(null)
  const open = Boolean(anchorEl)
  const handleClick = (event: React.SyntheticEvent<HTMLElement>) => {
    setAnchorEl(event.currentTarget)
  }
  const handleClose = () => {
    setAnchorEl(null)
  }

  return (
    <>
      <IconButton
        aria-label="more"
        id="long-button"
        aria-controls={open ? 'long-menu' : undefined}
        aria-expanded={open ? 'true' : undefined}
        aria-haspopup="true"
        onClick={handleClick}
        color="inherit"
        data-testid={dataTestId}
      >
        <MenuIcon />
      </IconButton>
      <Menu
        id="long-menu"
        MenuListProps={{
          'aria-labelledby': 'long-button',
        }}
        anchorEl={anchorEl}
        open={open}
        onClose={handleClose}
        PaperProps={{
          style: {
            maxHeight: ITEM_HEIGHT * 5,
            width: '20ch',
          },
        }}
      >
        {options.map(
          ({
            id,
            Icon,
            label,
            Component,
            'data-testid': optionDataTID,
            disabled,
          }: MenuOption) =>
            Component ? (
              <Component key={id} />
            ) : (
              <MenuItem
                key={id}
                selected={id === selected}
                onClick={() => {
                  setSelected(id)
                  onClick({ id, label })
                  handleClose()
                }}
                data-testid={optionDataTID}
                disabled={disabled}
              >
                {Icon && (
                  <ListItemIcon>
                    <Icon fontSize="small" />
                  </ListItemIcon>
                )}
                <ListItemText>{label}</ListItemText>
              </MenuItem>
            )
        )}
      </Menu>
    </>
  )
}
