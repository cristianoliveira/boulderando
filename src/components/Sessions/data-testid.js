// eslint-disable-next-line
export const SESSION_LIST_TABLE_ITEM = 'session_list_table_item'
