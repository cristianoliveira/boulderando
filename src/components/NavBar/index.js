import * as React from 'react'
import { useRouter } from 'next/router'

import AppBar from '@mui/material/AppBar'
import Toolbar from '@mui/material/Toolbar'
import Typography from '@mui/material/Typography'
import AccountCircle from '@mui/icons-material/AccountCircle'
import HistoryIcon from '@mui/icons-material/History'
import ListAltIcon from '@mui/icons-material/ListAlt'
import DeleteOutlineIcon from '@mui/icons-material/DeleteOutline'
import EditIcon from '@mui/icons-material/Edit'
import AddBoxIcon from '@mui/icons-material/AddBox'
import SyncIcon from '@mui/icons-material/Sync'

import MoreVertIcon from '@mui/icons-material/MoreVert'
import NavBarMenu from './NavBarMenu'

import useUserContext from '../../context/User'
import useSessionContext from '../../context/Sessions'

const PAGE_TITLES = {
  '/': 'Book Session',
  '/sessions': 'Book Session',
  '/user/new': 'New user',
  '/user/edit': 'Editing user',
}

export const NAVBAR_CONTAINER = 'navbar_container'

export const NAVBAR_USER_MENU_BUTTON = 'navbar_user_menu_button'
export const NAVBAR_USER_MENU_CONTAINER = 'navbar_user_menu_button--container'
export const NAVBAR_USER_MENU_ITEM_DELETE = 'navbar_user_menu_button--delete'
export const NAVBAR_USER_MENU_ITEM_EDIT = 'navbar_user_menu_button--edit'
export const NAVBAR_USER_MENU_ITEM_SYNC_DEVICE =
  'navbar_user_menu_item_sync_device'

export const NAVBAR_NAVIGATION_MENU_BUTTON = 'navbar_navigation_menu_button'
export const NAVBAR_NAVIGATION_MENU_ITEM_SESSION =
  'navbar_navigation_menu_item_session'
export const NAVBAR_NAVIGATION_MENU_ITEM_HISTORY =
  'navbar_navigation_menu_item_history'

export const NAVBAR_NAVIGATION_MENU_ITEM_CUSTOM_ADD =
  'navbar_navigation_menu_item_custom--add'

export const NAVBAR_NAVIGATION_MENU_ITEM_CUSTOM_DELETE =
  'navbar_navigation_menu_item_custom--delete'

export const NAVBAR_BOOKING_HISTORY_MENU = 'navbar_booking_history_button'

export default function NavBar() {
  const { user, deleteUser, editUser } = useUserContext()
  const { deleteCustomSessions } = useSessionContext()
  const router = useRouter()

  return (
    <AppBar position="static" data-testid={NAVBAR_CONTAINER}>
      <Toolbar>
        <NavBarMenu
          data-testid={NAVBAR_NAVIGATION_MENU_BUTTON}
          MenuIcon={MoreVertIcon}
          options={[
            {
              id: 'booking-history',
              Icon: HistoryIcon,
              label: 'History',
              'data-testid': NAVBAR_NAVIGATION_MENU_ITEM_HISTORY,
              disabled: !user,
            },
            {
              id: 'session',
              Icon: ListAltIcon,
              label: 'Sessions',
              'data-testid': NAVBAR_NAVIGATION_MENU_ITEM_SESSION,
              disabled: !user,
            },
            {
              id: 'add-custom-session',
              Icon: AddBoxIcon,
              label: 'Add Custom',
              'data-testid': NAVBAR_NAVIGATION_MENU_ITEM_CUSTOM_ADD,
              disabled: !user,
            },
            {
              id: 'delete-custom-session',
              Icon: DeleteOutlineIcon,
              label: 'Delete All',
              'data-testid': NAVBAR_NAVIGATION_MENU_ITEM_CUSTOM_DELETE,
              disabled: !user,
            },
            {
              id: 'sync-devices',
              Icon: SyncIcon,
              label: 'Sync Devices',
              'data-testid': NAVBAR_USER_MENU_ITEM_SYNC_DEVICE,
            },
          ]}
          onClick={({ id }) => {
            switch (id) {
              case 'session':
                router.push('/sessions')
                break

              case 'booking-history':
                router.push('/booking-history')
                break

              case 'add-custom-session':
                router.push('/sessions/new')
                break

              case 'delete-custom-session':
                deleteCustomSessions()
                break

              case 'sync-devices':
                router.push('/sync/devices')
                break

              default:
            }
          }}
        />
        <Typography variant="h6" component="div" sx={{ flexGrow: 1 }}>
          {PAGE_TITLES[router.asPath]}
        </Typography>
        {user && (
          <div>
            <NavBarMenu
              data-testid={NAVBAR_USER_MENU_BUTTON}
              MenuIcon={AccountCircle}
              options={[
                {
                  id: 'name',
                  label: user.name,
                  disabled: true,
                  'data-testid': NAVBAR_USER_MENU_CONTAINER,
                },
                {
                  id: 'delete',
                  Icon: DeleteOutlineIcon,
                  label: 'Delete',
                  'data-testid': NAVBAR_USER_MENU_ITEM_DELETE,
                },
                {
                  id: 'edit',
                  Icon: EditIcon,
                  label: 'Edit',
                  'data-testid': NAVBAR_USER_MENU_ITEM_EDIT,
                },
              ]}
              onClick={({ id }) => {
                switch (id) {
                  case 'delete':
                    deleteUser()
                    break

                  case 'edit':
                    editUser()
                    break

                  default:
                }
              }}
            />
          </div>
        )}
      </Toolbar>
    </AppBar>
  )
}
