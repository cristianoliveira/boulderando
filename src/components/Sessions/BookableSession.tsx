import React from 'react'
import { TableCell, TableCellProps, TableRow, Button } from '@mui/material'
import EventAvailableIcon from '@mui/icons-material/EventAvailable'

import getNextPossibleDay from '../../modules/weekday'

import { SESSION_LIST_TABLE_ITEM } from './data-testid'

const TbCell = (props: TableCellProps) => (
  <TableCell
    sx={{
      padding: '8px',
    }}
    {...props}
  />
)

type BookableSessionProps = {
  session: Session
  hasBeenBooked: boolean
  isProcessing: boolean
  onBookingSubmit(sessions: Session): void
}

const BookableSession: React.FC<BookableSessionProps> = ({
  session,
  hasBeenBooked,
  isProcessing,
  onBookingSubmit,
}) => (
  <TableRow
    data-testid={SESSION_LIST_TABLE_ITEM}
    sx={{
      color: hasBeenBooked ? 'gray' : 'green',
    }}
  >
    <TbCell>{session.gym_name}</TbCell>
    <TbCell>
      {session.day_of_week}-{getNextPossibleDay(session.day_of_week)}
    </TbCell>
    <TbCell>{session.time}</TbCell>
    <TbCell align={'right'}>
      <Button
        disabled={isProcessing || hasBeenBooked}
        variant="contained"
        type="submit"
        onClick={() => onBookingSubmit(session)}
      >
        <EventAvailableIcon />
      </Button>
    </TbCell>
  </TableRow>
)

export default BookableSession
