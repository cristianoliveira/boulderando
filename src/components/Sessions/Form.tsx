import styled from '@emotion/styled'

import FormControl from '@mui/material/FormControl'
import FormHelperText from '@mui/material/FormHelperText'
import Button from '@mui/material/Button'
import Grid from '@mui/material/Grid'
import { useForm } from 'react-hook-form'
import Alert from '@mui/material/Alert'

import SelectField from '../fields/SelectField'

import * as TID from '../../constants/data-testid'

const StyledForm = styled.form`
  padding: 20px;
`

const FULL_GRID = 12

const showErrorFields = (errors: object): string =>
  Object.keys(errors)
    .map((e) => e)
    .join(', ')

// Sometime in life you have to just have faith
// range from 8:00 to 21:30
const timesSelection: Array<string> = [...Array(20).keys()]
  .filter((i) => i >= 8 && i <= 20)
  .map((i) => [`${i}:00 - ${i + 2}:00`, `${i}:30 - ${i + 2}:30`])
  .flat()

function Form({ onSubmit }: Submitable): JSX.Element {
  const {
    register,
    handleSubmit,
    formState: { errors, isValid, isSubmitted },
  } = useForm({})

  const errorsList = showErrorFields(errors)
  return (
    <StyledForm
      onSubmit={handleSubmit((data) =>
        onSubmit({ ...data, day_of_week: data.human_date.replace(/this /, '') })
      )}
    >
      <Grid
        direction="row"
        alignItems="flex-start"
        spacing={2}
        container
      >
        <Grid item xs={FULL_GRID}>
          <SelectField
            data-testid={TID.CUSTOM_SESSION_FORM_GYM}
            name="gym_name"
            label="Gym"
            register={register}
            selectOptions={[
              { label: 'Boulderklub', value: 'boulderklub' },
              { label: 'Basement', value: 'basement' },
            ]}
          />
        </Grid>
        <Grid item xs={FULL_GRID}>
          <SelectField
            data-testid={TID.CUSTOM_SESSION_FORM_DAY}
            name="human_date"
            label="Day of the week"
            register={register}
            selectOptions={[
              { label: 'Monday', value: 'this monday' },
              { label: 'Tuesday', value: 'this tuesday' },
              { label: 'Wednesday', value: 'this wednesday' },
              { label: 'Thursday', value: 'this thursday' },
              { label: 'Friday', value: 'this friday' },
              { label: 'Saturday', value: 'this saturday' },
            ]}
          />
        </Grid>
        <Grid item xs={FULL_GRID}>
          <SelectField
            data-testid={TID.CUSTOM_SESSION_FORM_TIME}
            name="time"
            label="Time"
            register={register}
            selectOptions={timesSelection.map((time) => ({
              label: time,
              value: time,
            }))}
          />
        </Grid>
        <Grid item xs={FULL_GRID}>
          {isSubmitted && !isValid && errorsList && (
            <Alert
              data-testid={TID.USER_FORM_ERROR_MESSAGE_CONTAINER}
              severity="error"
            >
              All fields are required. Missing fields: {errorsList}
            </Alert>
          )}
        </Grid>
      </Grid>
      <FormControl fullWidth>
        <Button
          data-testid={TID.CUSTOM_SESSION_FORM_SUBMIT_BUTTON}
          variant="contained"
          type="submit"
        >
          Save data
        </Button>
      </FormControl>
      <FormHelperText variant="outlined">
        ** This data is stored in your browser and can be deleted at any time
      </FormHelperText>
    </StyledForm>
  )
}

export default Form
