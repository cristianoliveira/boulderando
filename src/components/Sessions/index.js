import React from 'react'
import { SessionProvider, SessionConsumer } from '../../context/Sessions'
import SessionList from './SessionList'

function Session({ user }) {
  return (
    <SessionProvider user={user}>
      <SessionConsumer>
        {(context) => (
          <SessionList {...context} />
        )}
      </SessionConsumer>
    </SessionProvider>
  )
}

export default Session
