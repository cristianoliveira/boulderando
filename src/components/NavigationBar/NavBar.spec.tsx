import React from 'react'
// eslint-disable-next-line
import renderer from 'react-test-renderer'

/* eslint-disable */
import { UserProvider } from '../../../src/context/User'
import { SessionProvider } from '../../../src/context/Sessions'
/* eslint-enable */

import NavBar from './NavBar'

describe('NavBar', () => {
  it('matches the snapshot', async () => {
    expect(
      renderer
        .create(
          <UserProvider api={undefined as any}>
            <SessionProvider api={undefined as any}>
              <NavBar pageTitle="MyPaget Title" />
            </SessionProvider>
          </UserProvider>
        )
        .toJSON()
    ).toMatchSnapshot()
  })
})
