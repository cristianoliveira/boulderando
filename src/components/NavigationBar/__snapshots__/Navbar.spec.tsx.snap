// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar matches the snapshot 1`] = `
<header
  className="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-hip9hq-MuiPaper-root-MuiAppBar-root"
  data-testid="navbar_container"
>
  <div
    className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1t29gy6-MuiToolbar-root"
  >
    <button
      aria-haspopup="true"
      aria-label="more"
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
      data-testid="navbar_navigation_menu_button"
      disabled={false}
      id="long-button"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="MoreVertIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
        />
      </svg>
    </button>
    <div
      className="MuiTypography-root MuiTypography-h6 css-1juivf6-MuiTypography-root"
    >
      MyPaget Title
    </div>
  </div>
</header>
`;
